### ISSUE TITLE
Improve Google Drive performance using gzip

### ISSUE THREAD

**Issue by Dibbyo456** [rclone/rclone#3870]
According to Google Drive API docs, [Google recommends](https://developers.google.com/drive/api/v3/performance#gzip) to use `gzip` compression when sending requests.

In order to receive a gzip-encoded response you **must do two things**: Set an  `Accept-Encoding`  header, and modify your user agent to contain the string  `gzip` . Here is an example of properly formed HTTP headers for enabling gzip compression:

```
Accept-Encoding: gzip
User-Agent: my program (gzip)
```

Currently Rclone send GET request using something like this:
```
2020/01/12 12:11:11 DEBUG : GET /drive/v3/files?{{redacted}}&supportsAllDrives=true HTTP/1.1
Host: www.googleapis.com
User-Agent: rclone/v1.50.2
Authorization: XXXX
X-Goog-Api-Client: gl-go/1.11.0 gdcl/20190802
Accept-Encoding: gzip
```
The `Accept-Encoding: gzip` is available to request headers but string `gzip` is missing from `User-Agent`.

According to Google, Rclone should send:
```
User-Agent: rclone/v1.50.2 (gzip)
```
in order to receive a gzip-encoded response.

**Comment by ncw** [comment]
Can you try adding `--user-agent "rclone/v1.50.2 (gzip)"` and see if

- the response is gzipped
- rclone carries on working

If you could do some performance tests that would be useful too.

**Comment by Dibbyo456** [comment]
> Can you try adding `--user-agent "rclone/v1.50.2 (gzip)"`

This is weird. Google drive api does not send gzipped data even when the `Accept-Encoding: gzip` and `--user-agent "rclone/v1.50.2 (gzip)"` is present on Rclone. 😕

```
2020/01/12 20:38:51 DEBUG : HTTP REQUEST (req 0xc000340100)
2020/01/12 20:38:51 DEBUG : GET /drive/v3/about?alt=json&fields=exportFormats%2CimportFormats&prettyPrint=false HTTP/1.1
Host: www.googleapis.com
User-Agent: rclone/v1.50.2 (gzip)
Authorization: XXXX
X-Goog-Api-Client: gl-go/1.11.0 gdcl/20190802
Accept-Encoding: gzip

>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<

2020/01/12 20:38:52 DEBUG : HTTP RESPONSE (req 0xc000340100)
2020/01/12 20:38:52 DEBUG : HTTP/1.1 200 OK
Transfer-Encoding: chunked
Alt-Svc: quic=":443"; ma=2592000; v="46,43",h3-Q050=":443"; ma=2592000,h3-Q049=":443"; ma=2592000,h3-Q048=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000
Cache-Control: private, max-age=0, must-revalidate, no-transform
Content-Type: application/json; charset=UTF-8
Date: Sun, 12 Jan 2020 15:08:52 GMT
Expires: Sun, 12 Jan 2020 15:08:52 GMT
Server: GSE
Vary: Origin
Vary: X-Origin
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Xss-Protection: 1; mode=block
```
Someone had the same issue on [stackoverflow](https://stackoverflow.com/questions/32004933/download-file-from-google-drive-via-rest-api-using-gzip-compression).

I did some more testing using `curl`.

```
curl -v -H "Accept-Encoding: gzip" -H "Authorization: Bearer {{redacted}}" "https://www.googleapis.com/drive/v3/about?alt=json&fields=storageQuota&prettyPrint=false"
```

**Verbose Output**:

```
>>>>>>>>>>>>>>>>>>>>>>>>>>>
>>>>>>>>>>>>>>>>>>>>>>>>>>>

GET /drive/v3/about?alt=json&fields=storageQuota&prettyPrint=false HTTP/1.1
Host: www.googleapis.com
User-Agent: curl/7.55.1
Accept: */*
Accept-Encoding: gzip
Authorization: Bearer {{redacted}}

<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
<<<<<<<<<<<<<<<<<<<<<<<<<<<<<

HTTP/1.1 200 OK
Expires: Sun, 12 Jan 2020 16:53:23 GMT
Date: Sun, 12 Jan 2020 16:53:23 GMT
Cache-Control: private, max-age=0, must-revalidate, no-transform
Vary: Origin
Vary: X-Origin
Content-Type: application/json; charset=UTF-8
Content-Encoding: gzip
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-XSS-Protection: 1; mode=block
Server: GSE
Alt-Svc: quic=":443"; ma=2592000; v="46,43",h3-Q050=":443"; ma=2592000,h3-Q049=":443"; ma=2592000,h3-Q048=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000
Transfer-Encoding: chunked
```
Surprisingly `Content-Encoding: gzip` is present on the response header.
So, basically when using `curl` it works, (even without the `gzip` string on the user-agent).

**Comment by ncw** [comment]
Note that the go HTTP library does automatic gzip decompression so looking at the results of `-vv --dump headers` is probably always going to show the decompressed data. So we might need a better way of determining whether the data is compressed or not.  Maybe looking at the total sizes of the transfer in Wireshark?

I uploaded two files one full of nulls 10M-1 and one full of random data 10M-urandom, both 10M in size.

Here are the results of me downloading it on my home internet connection which maxes out at about 5MBytes/s.

Judging from those I'd say rclone is already using gzip encoding. with drive regardless of the state of the user agent.

However I tried the same thing with a 1GB file of zeros and it downloaded at the same speed regardless which is confusing.

## random file, no gzip user agent

```
$ rclone copyto -P --user-agent "rclone/v1.50.2 (null)" TestDrive:10M-urandom  /tmp/10M-1-urandome
Transferred:           10M / 10 MBytes, 100%, 2.518 MBytes/s, ETA 0s
Transferred:            1 / 1, 100%
Elapsed time:        3.9s
```

## random file, gzip user agent

```
$ rclone copyto -P --user-agent "rclone/v1.50.2 (gzip)" TestDrive:10M-urandom  /tmp/10M-1-urandom-gzip
Transferred:           10M / 10 MBytes, 100%, 2.229 MBytes/s, ETA 0s
Transferred:            1 / 1, 100%
Elapsed time:        4.4s
```

## zeros file no gzip user agent

```
$ rclone copyto -P --user-agent "rclone/v1.50.2 (null)" TestDrive:10M-1  /tmp/10M-1-null
Transferred:           10M / 10 MBytes, 100%, 22.618 MBytes/s, ETA 0s
Transferred:            1 / 1, 100%
Elapsed time:       400ms
```

## zeros file, gzip user agent

```
$ rclone copyto -P --user-agent "rclone/v1.50.2 (gzip)" TestDrive:10M-1  /tmp/10M-1-gzip
Transferred:           10M / 10 MBytes, 100%, 14.579 MBytes/s, ETA 0s
Transferred:            1 / 1, 100%
Elapsed time:       600ms
```

**Comment by Dibbyo456** [comment]
> Note that the go HTTP library does automatic gzip decompression so looking at the results of `-vv --dump headers` is probably always going to show the decompressed data.

But then why isn't there any `Content-Encoding: gzip` on the response headers? There is no way rclone or the go http library suppressing the header from response.

> Judging from those I'd say rclone is already using gzip encoding. with drive regardless of the state of the user agent.

True. I have tested the same with `curl`, with/without the `gzip` string on user-agent, google drive sends gzipped data.

But when using `curl` the `Content-Encoding: gzip` is present on the response headers, where Rclone it's missing.

**Comment by Dibbyo456** [comment]
> > Note that the go HTTP library does automatic gzip decompression so looking at the results of `-vv --dump headers` is probably always going to show the decompressed data.
> 
> But then why isn't there any `Content-Encoding: gzip` on the response headers? There is no way rclone or the go http library suppressing the header from response.

hahaha.. @ncw Surprisingly go http library seems to suppress the `Content-Encoding: gzip` headers from response. 😑

Read the **Steffen Ullrich**'s comment from the [issue on stackoverflow](https://stackoverflow.com/a/59125849/2277053).

Closing the issue for now, feel free to open if you think otherwise.

**Comment by ncw** [comment]
Yes I think you are right.

I also managed to demonstrate downloading the 1GB file with compression by turning off multi thread copy. I watched the network speed while doing this - at no point did it go above 1.5 MB/s!

```
$ rclone copyto -v --multi-thread-streams 0 TestDrive:1GB-zeros  /tmp/1GB-zeros3
Transferred:   	        1G / 1 GBytes, 100%, 204.697 MBytes/s, ETA 0s
Transferred:            1 / 1, 100%
Elapsed time:          5s
```

The fact this didn't work with multi thread streams probably means that google doesn't do gzip compression if you supply a Range header.

So I think case closed!
